community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntity.java

```diff
@@ -19,204 +19,373 @@
-            {
-                free = i - 1;
-                break;
-            }
-            if ( existingKey.equals( key ) )
-            {   // update
-                properties[i] = behaviour.merge( properties[i], value );
-                return;
-            }
-        }
+        checkClear();
+        hasLabelField = true;
+        this.labelField = labelField;
+        return delegate.labelField( labelField );
+    }
 
-        // Add
-        properties[free++] = key;
-        properties[free] = value;
+    @Override
+    public boolean startId( long id )
+    {
+        checkClear();
+        hasLongStartId = true;
+        longStartId = id;
+        return delegate.startId( id );
     }
 
-    public void setProperties( Object... keyValuePairs )
+    @Override
+    public boolean startId( Object id, Group group )
     {
-        properties = keyValuePairs;
-        firstPropertyId = null;
+        checkClear();
+        objectStartId = id;
+        startIdGroup = group;
+        return delegate.startId( id, group );
     }
 
-    public boolean hasFirstPropertyId()
+    @Override
+    public boolean endId( long id )
     {
-        return firstPropertyId != null;
+        checkClear();
+        hasLongEndId = true;
+        longEndId = id;
+        return delegate.endId( id );
     }
 
-    public long firstPropertyId()
+    @Override
+    public boolean endId( Object id, Group group )
     {
-        return firstPropertyId;
+        checkClear();
+        objectEndId = id;
+        endIdGroup = group;
+        return delegate.endId( id, group );
     }
 
     @Override
-    public String sourceDescription()
+    public boolean type( int type )
     {
-        return sourceDescription;
+        checkClear();
+        hasIntType = true;
+        intType = type;
+        return delegate.type( type );
     }
 
     @Override
-    public long lineNumber()
+    public boolean type( String type )
     {
-        return lineNumber;
+        checkClear();
+        stringType = type;
+        return delegate.type( type );
     }
 
     @Override
-    public long position()
+    public void endOfEntity() throws IOException
     {
-        return position;
+        // Mark that the next call to any data method should clear the state
+        end = true;
+        delegate.endOfEntity();
     }
 
-    public void rebase( long baseLineNumber, long basePosition )
+    public String[] labels()
     {
-        lineNumber += baseLineNumber;
-        position += basePosition;
+        return labels.toArray( new String[labels.size()] );
     }
 
-    @Override
-    public String toString()
+    public Object[] properties()
+    {
+        return properties.toArray();
+    }
+
+    public Object id()
     {
-        Collection<Pair<String,?>> fields = new ArrayList<>();
-        toStringFields( fields );
+        return hasLongId ? longId : objectId;
+    }
+
+    public Object endId()
+    {
+        return hasLongEndId ? longEndId : objectEndId;
```

Comment:
Could the auto-box be avoided here? Or will exposing getters for each of the `has*`, `long*` and `object*` fields just move the same auto-boxing further up the stack?

Reply:
I think it would unfortunately. This isn't part of the main paths anyway.